Mathlib/Topology/Compactness/Lindelof.lean

```diff
@@ -0,0 +1,357 @@
+        have : ⋃ s∈ Se, s = ht₁ᶜ ∪ ht₂ᶜ := by simp
+        rwa [← this]
+        }
+  have : CountableInterFilter f := by
+    apply CountableInterFilter.mk
+    simp only [Filter.mem_mk, mem_setOf_eq, compl_sInter, sUnion_image]
+    intro S hS hsp
+    let f := fun (x : Set X) ↦ xᶜ
+    let S' := f '' S
+    have hsp : ∀ s ∈ S', p s := by simpa
+    have hS' : S'.Countable := Countable.image hS _
+    have : ⋃ s ∈ S, sᶜ = ⋃ s ∈ S', s := by simp
+    rw [this]
+    apply hcountable_union S' hS' hsp
+  have : sᶜ ∈ f := hs.compl_mem_sets_of_nhdsWithin (by simpa using hnhds)
+  rwa [← compl_compl s]
+
+/-- The intersection of a Lindelöf set and a closed set is a Lindelöf set. -/
+theorem IsLindelof.inter_right (hs : IsLindelof s) (ht : IsClosed t) : IsLindelof (s ∩ t) := by
+  intro f hnf _ hstf
+  obtain ⟨x, hsx, hx⟩ : ∃ x ∈ s, ClusterPt x f :=
+    hs (le_trans hstf (le_principal_iff.2 (inter_subset_left _ _)))
+  have : x ∈ t := ht.mem_of_nhdsWithin_neBot <|
+    hx.mono <| le_trans hstf (le_principal_iff.2 (inter_subset_right _ _))
+  exact ⟨x, ⟨hsx, this⟩, hx⟩
+
+  /-- The intersection of a closed set and a Lindelöf set is a Lindelöf set. -/
+theorem IsLindelof.inter_left (ht : IsLindelof t) (hs : IsClosed s) : IsLindelof (s ∩ t) :=
+  inter_comm t s ▸ ht.inter_right hs
+
+  /-- The set difference of a Lindelöf set and an open set is a Lindelöf set. -/
+theorem IsLindelof.diff (hs : IsLindelof s) (ht : IsOpen t) : IsLindelof (s \ t) :=
+  hs.inter_right (isClosed_compl_iff.mpr ht)
+
+/-- A closed subset of a Lindelöf set is a Lindelöf set. -/
+theorem IsLindelof.of_isClosed_subset (hs : IsLindelof s) (ht : IsClosed t) (h : t ⊆ s) :
+    IsLindelof t := inter_eq_self_of_subset_right h ▸ hs.inter_right ht
+
+/-- A continuous image of a Lindelöf set is a Lindelöf set. -/
+theorem IsLindelof.image_of_continuousOn {f : X → Y} (hs : IsLindelof s) (hf : ContinuousOn f s) :
+    IsLindelof (f '' s) := by
+  intro l lne _ ls
+  have : NeBot (l.comap f ⊓ 𝓟 s) :=
+    comap_inf_principal_neBot_of_image_mem lne (le_principal_iff.1 ls)
+  obtain ⟨x, hxs, hx⟩ : ∃ x ∈ s, ClusterPt x (l.comap f ⊓ 𝓟 s) := @hs _ this _ inf_le_right
+  haveI := hx.neBot
+  use f x, mem_image_of_mem f hxs
+  have : Tendsto f (𝓝 x ⊓ (comap f l ⊓ 𝓟 s)) (𝓝 (f x) ⊓ l) := by
+    convert (hf x hxs).inf (@tendsto_comap _ _ f l) using 1
+    rw [nhdsWithin]
+    ac_rfl
+  exact this.neBot
+
+/-- A continuous image of a Lindelöf set is a Lindelöf set within the codomain. -/
+theorem IsLindelof.image {f : X → Y} (hs : IsLindelof s) (hf : Continuous f) :
+    IsLindelof (f '' s) := hs.image_of_continuousOn hf.continuousOn
+
+/-- A filter with the countable intersection property that is finer than the principal filter on
+a Lindelöf set `s` contains any open set that contains all clusterpoints of `s`. -/
+theorem IsLindelof.adherence_nhdset {f : Filter X} [CountableInterFilter f] (hs : IsLindelof s)
+    (hf₂ : f ≤ 𝓟 s) (ht₁ : IsOpen t) (ht₂ : ∀ x ∈ s, ClusterPt x f → x ∈ t) : t ∈ f :=
```

Comment:
While lots of API about compact sets is written with `f ≤ 𝓟 s` assumptions here and there, I think that having assumptions in non-simp form is not a good design decision. I would assume `s ∈ f` instead. But AFAIK we don't have coding style guidelines for this.

Reply:
I agree that 's in f' may be more elegant, I just tried to keep as close to Compact as possible. Would you like me to change these? I'll make a PR with the other changes in the meanwhile, I'm happy to make a new one with this specific change!